File: iOS/Kalorie/Features/Moderation/ModerationQueueViewModel.swift
//
//  ModerationQueueViewModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 10.09.2026.
//

import Foundation

final class ModerationQueueViewModel: ObservableObject {

    // MARK: - Properties

    @Published private(set) var state: LoadingState<Void> = .idle
    @Published private(set) var submissions: [FoodItemSubmissionDomain] = []
    @Published private(set) var collidingBarcodes: Set<String> = []
    @Published var alertItem: AlertItem?

    private let fetchPendingSubmissions: any FetchPendingSubmissionsUseCaseProtocol
    private let fetchFoodItemByBarcode: any FetchFoodItemByBarcodeUseCaseProtocol

    // MARK: - Init

    init(
        fetchPendingSubmissions: any FetchPendingSubmissionsUseCaseProtocol,
        fetchFoodItemByBarcode: any FetchFoodItemByBarcodeUseCaseProtocol
    ) {
        self.fetchPendingSubmissions = fetchPendingSubmissions
        self.fetchFoodItemByBarcode = fetchFoodItemByBarcode
    }

    // MARK: - Functions

    @MainActor
    func onAppear() async {
        state = .loading
        defer { state = .loaded }
        do {
            submissions = try await fetchPendingSubmissions()
            await refreshCollisions()
        } catch {
            Log.error(error, category: Constants.LogCategory.moderation)
            alertItem = AlertItem(title: L10n.Common.errorUnknown)
        }
    }

    @MainActor
    func onRefresh() async {
        await onAppear()
    }

    @MainActor
    func onSubmissionResolved(id: String) async {
        submissions.removeAll { $0.id == id }
        await refreshCollisions()
    }

    func isColliding(_ submission: FoodItemSubmissionDomain) -> Bool {
        guard let barcode = submission.barcode else { return false }
        return collidingBarcodes.contains(barcode)
    }

    // MARK: - Private

    @MainActor
    private func refreshCollisions() async {
        let barcodes = submissions.compactMap(\.barcode)
        collidingBarcodes = await Self.collidingBarcodes(among: barcodes, fetchFoodItemByBarcode: fetchFoodItemByBarcode)
    }

    private static func collidingBarcodes(
        among barcodes: [String],
        fetchFoodItemByBarcode: any FetchFoodItemByBarcodeUseCaseProtocol
    ) async -> Set<String> {
        let existing = (try? await fetchFoodItemByBarcode(barcodes: barcodes)) ?? []
        return Set(existing.map(\.id))
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/moderation/ModerationQueueViewModel.kt
package antoni.kalorie.features.moderation

import androidx.lifecycle.ViewModel
import antoni.kalorie.R
import antoni.kalorie.core.models.FoodItemSubmissionDomain
import antoni.kalorie.core.usecases.FetchFoodItemByBarcodeUseCaseProtocol
import antoni.kalorie.core.usecases.FetchPendingSubmissionsUseCaseProtocol
import antoni.kalorie.core.utils.AlertItem
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.LoadingState
import antoni.kalorie.core.utils.Log
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class ModerationQueueViewModel(
    private val fetchPendingSubmissions: FetchPendingSubmissionsUseCaseProtocol,
    private val fetchFoodItemByBarcode: FetchFoodItemByBarcodeUseCaseProtocol,
) : ViewModel() {

    // MARK: - Properties

    private val _state = MutableStateFlow<LoadingState<Unit>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<Unit>> = _state
    private val _submissions = MutableStateFlow<List<FoodItemSubmissionDomain>>(emptyList())
    val submissions: StateFlow<List<FoodItemSubmissionDomain>> = _submissions
    private val _collidingBarcodes = MutableStateFlow<Set<String>>(emptySet())
    val collidingBarcodes: StateFlow<Set<String>> = _collidingBarcodes
    val alertItem = MutableStateFlow<AlertItem?>(null)

    // MARK: - Functions

    suspend fun onAppear() {
        _state.value = LoadingState.Loading
        try {
            _submissions.value = fetchPendingSubmissions()
            refreshCollisions()
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.MODERATION)
            alertItem.value = AlertItem(titleRes = R.string.common_error_unknown)
        } finally {
            _state.value = LoadingState.loaded
        }
    }

    suspend fun onRefresh() {
        onAppear()
    }

    suspend fun onSubmissionResolved(id: String) {
        _submissions.value = _submissions.value.filterNot { it.id == id }
        refreshCollisions()
    }

    fun isColliding(submission: FoodItemSubmissionDomain): Boolean {
        val barcode = submission.barcode ?: return false
        return barcode in _collidingBarcodes.value
    }

    // MARK: - Private

    private suspend fun refreshCollisions() {
        val barcodes = _submissions.value.mapNotNull { it.barcode }
        val existing = try {
            fetchFoodItemByBarcode(barcodes)
        } catch (error: CancellationException) {
            throw error
        } catch (_: Exception) {
            emptyList()
        }
        _collidingBarcodes.value = existing.map { it.id }.toSet()
    }
}
